### PR TITLE
Mirage Pipeline

### PR DESCRIPTION
**Pull request by DavidBert**
# Mirage pipeline

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: scripts/convert_mirage_to_diffusers.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+from diffusers.pipelines.mirage import MiragePipeline
+
+
+def load_reference_config(vae_type: str) -> dict:
+    """Load transformer config from existing pipeline checkpoint."""
+
+    if vae_type == "flux":
+        config_path = "/raid/shared/storage/home/davidb/diffusers/diffusers_pipeline_checkpoints/pipeline_checkpoint_fluxvae_gemmaT5_updated/transformer/config.json"
+    elif vae_type == "dc-ae":
+        config_path = "/raid/shared/storage/home/davidb/diffusers/diffusers_pipeline_checkpoints/pipeline_checkpoint_dcae_gemmaT5_updated/transformer/config.json"
```
Comment (photoroman) [inline scripts/convert_mirage_to_diffusers.py]: Should we change these hardcoded paths?

File: scripts/convert_photon_to_diffusers.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
+#!/usr/bin/env python3
+"""
+Script to convert Mirage checkpoint from original codebase to diffusers format.
+"""
```
Comment (photoroman) [inline scripts/convert_photon_to_diffusers.py:4]: Should we release this script or put it into the computer vision repo and release the converted checkpoints?

File: scripts/convert_mirage_to_diffusers.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+
+
+def create_parameter_mapping() -> dict:
+    """Create mapping from old parameter names to new diffusers names."""
+
+    # Key mappings for structural changes
+    mapping = {}
+
+    # RMSNorm: scale -> weight
+    for i in range(16):  # 16 layers
```
Comment (photoroman) [inline scripts/convert_mirage_to_diffusers.py]: Use a `MIRAGE_NUM_LAYERS: int = 16` constant at the top?

File: scripts/convert_mirage_to_diffusers.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+    return transformer
+
+
+def copy_pipeline_components(vae_type: str, output_path: str):
+    """Copy VAE, scheduler, text encoder, and tokenizer from reference pipeline."""
+
+    if vae_type == "flux":
+        ref_pipeline = "/raid/shared/storage/home/davidb/diffusers/diffusers_pipeline_checkpoints/pipeline_checkpoint_fluxvae_gemmaT5_updated"
+    else:  # dc-ae
+        ref_pipeline = "/raid/shared/storage/home/davidb/diffusers/diffusers_pipeline_checkpoints/pipeline_checkpoint_dcae_gemmaT5_updated"
```
Comment (photoroman) [inline scripts/convert_mirage_to_diffusers.py]: Should we change these hardcoded paths?

File: scripts/convert_mirage_to_diffusers.py [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+        mapping[f"blocks.{i}.qk_norm.key_norm.scale"] = f"blocks.{i}.qk_norm.key_norm.weight"
+        mapping[f"blocks.{i}.k_norm.scale"] = f"blocks.{i}.k_norm.weight"
+
+        # Attention: attn_out -> attention.to_out.0
+        mapping[f"blocks.{i}.attn_out.weight"] = f"blocks.{i}.attention.to_out.0.weight"
+
+    return mapping
+
+
+def convert_checkpoint_parameters(old_state_dict: dict) -> dict:
```
Comment (photoroman) [inline scripts/convert_mirage_to_diffusers.py]: Probably good to use more specific types, like `Dict[str, str]`. Import `Dict` from `typing`, because the `diffusers` library supports Python 3.8 and built-in types with generics, e.g. `dict[str, str]` are only supported since Python 3.9.

File: src/diffusers/models/transformers/transformer_photon.py [hunk 6]
```diff
@@ -0,0 +1,502 @@
...
+
+
+def apply_rope(xq: Tensor, freqs_cis: Tensor) -> Tensor:
+    xq_ = xq.float().reshape(*xq.shape[:-1], -1, 1, 2)
+    xq_out = freqs_cis[..., 0] * xq_[..., 0] + freqs_cis[..., 1] * xq_[..., 1]
+    return xq_out.reshape(*xq.shape).type_as(xq)
+
+
+class EmbedND(nn.Module):
+    def __init__(self, dim: int, theta: int, axes_dim: list[int]):
```
Comment (photoroman) [inline src/diffusers/models/transformers/transformer_photon.py:101]: Use `List[int]` for Python 3.8 support.

File: src/diffusers/models/transformers/transformer_mirage.py [hunk 7]
```diff
@@ -0,0 +1,502 @@
...
+from ..embeddings import get_timestep_embedding
+from ..modeling_outputs import Transformer2DModelOutput
+from ..modeling_utils import ModelMixin
+from ..normalization import RMSNorm
+
+
+logger = logging.get_logger(__name__)
+
+
+def get_image_ids(bs: int, h: int, w: int, patch_size: int, device: torch.device) -> Tensor:
```
Comment (photoroman) [inline src/diffusers/models/transformers/transformer_mirage.py]: - Use more readable variable names: `batch_size`, `height`, `width`.
- Add docstrings for all public methods and classes.

File: src/diffusers/models/transformers/transformer_mirage.py [hunk 8]
```diff
@@ -0,0 +1,502 @@
...
+        # build multiplicative 0/1 mask for provided attention_mask over [cond?, text, image] keys
+        attn_mask: Tensor | None = None
+        if attention_mask is not None:
+            bs, _, l_img, _ = img_q.shape
+            l_txt = txt_k.shape[2]
+
+            assert attention_mask.dim() == 2, f"Unsupported attention_mask shape: {attention_mask.shape}"
+            assert attention_mask.shape[-1] == l_txt, (
+                f"attention_mask last dim {attention_mask.shape[-1]} must equal text length {l_txt}"
+            )
```
Comment (photoroman) [inline src/diffusers/models/transformers/transformer_mirage.py]: Should these be checked as `if` conditions and raised as `ValueError`? Usually asserts are to catch programming errors not for input validation.

File: src/diffusers/pipelines/photon/pipeline_photon.py [hunk 9]
```diff
@@ -0,0 +1,651 @@
...
+        scheduler ([`FlowMatchEulerDiscreteScheduler`]):
+            A scheduler to be used in combination with `transformer` to denoise the encoded image latents.
+        text_encoder ([`T5EncoderModel`]):
+            Standard text encoder model for encoding prompts.
+        tokenizer ([`T5TokenizerFast` or `GemmaTokenizerFast`]):
+            Tokenizer for the text encoder.
+        vae ([`AutoencoderKL`] or [`AutoencoderDC`]):
+            Variational Auto-Encoder (VAE) Model to encode and decode images to and from latent representations.
+            Supports both AutoencoderKL (8x compression) and AutoencoderDC (32x compression).
+    """
```
Comment (photoroman) [inline src/diffusers/pipelines/photon/pipeline_photon.py:218]: I think in addition to this, we need to add docs about Mirage in `docs/source/en/api/pipelines/mirage.md`. Have a look at the section "Adding a new pipeline/scheduler" in `docs/README.md`.

File: src/diffusers/pipelines/mirage/pipeline_mirage.py [hunk 10]
```diff
@@ -0,0 +1,651 @@
...
+
+        Returns:
+            [`~pipelines.mirage.MiragePipelineOutput`] or `tuple`: [`~pipelines.mirage.MiragePipelineOutput`] if
+            `return_dict` is True, otherwise a `tuple. When returning a tuple, the first element is a list with the
+            generated images.
+        """
+
+        # 0. Default height and width to transformer config
+        height = height or 256
+        width = width or 256
```
Comment (photoroman) [inline src/diffusers/pipelines/mirage/pipeline_mirage.py]: nit: put 256 into a constant.

File: src/diffusers/pipelines/mirage/pipeline_mirage.py [hunk 11]
```diff
@@ -0,0 +1,651 @@
...
+                img_seq = self.transformer.forward_transformers(
+                    img_seq,
+                    txt,
+                    time_embedding=self.transformer.compute_timestep_embedding(t_cont, img_seq.dtype),
+                    pe=pe,
+                    attention_mask=ca_mask,
+                )
+
+                # Convert back to image format
+                from ...models.transformers.transformer_mirage import seq2img
```
Comment (photoroman) [inline src/diffusers/pipelines/mirage/pipeline_mirage.py]: Any reason not to import this at the top?

**Review by photoroman — COMMENTED** [review]
I did a brief initial pass. This looks really great! Thank you!

The main missing part for me are more docs. I think we should add docstrings to all public classes and methods. Also we should add some docs and examples in `docs/source/en/api/pipelines/mirage.md`, probably.

File: src/diffusers/pipelines/mirage/pipeline_mirage.py (hunk 11 above)
Comment (almazan) [reply src/diffusers/pipelines/mirage/pipeline_mirage.py]: I've been becoming quite keen of doing in-code imports for cases like this one, where `seq2img` is only used once in the whole file :D

File: src/diffusers/pipelines/mirage/pipeline_mirage.py [hunk 12]
```diff
@@ -0,0 +1,651 @@
...
+        ... )
+        >>> pipe.to("cuda")
+        >>> prompt = "A digital painting of a rusty, vintage tram on a sandy beach"
+        >>> image = pipe(prompt, num_inference_steps=28, guidance_scale=4.0).images[0]
+        >>> image.save("mirage_output.png")
+        ```
+"""
+
+
+class MiragePipeline(
```
Comment (almazan) [inline src/diffusers/pipelines/mirage/pipeline_mirage.py]: I remember Eliot started a thread about maybe changing the name of Mirage to something else, since there's already a ML model called like that. Did that end up on something?

File: scripts/convert_photon_to_diffusers.py (hunk 2 above)
Comment (almazan) [reply scripts/convert_photon_to_diffusers.py:4]: I think the idea is to keep this only internally, right?

Comment (DavidBert) [reply scripts/convert_photon_to_diffusers.py:4]: Looking at other scripts in this folder, it seems that most companies actually put these kinds of scripts in diffusers.

File: scripts/convert_mirage_to_diffusers.py (hunk 1 above)
Comment (DavidBert) [reply scripts/convert_mirage_to_diffusers.py]: Absolutely!

File: src/diffusers/pipelines/mirage/pipeline_mirage.py (hunk 10 above)
Comment (DavidBert) [reply src/diffusers/pipelines/mirage/pipeline_mirage.py]: I specify it in the checkpoint config now and changed it for a constant when not specified

File: src/diffusers/pipelines/mirage/pipeline_mirage.py (hunk 11 above)
Comment (DavidBert) [reply src/diffusers/pipelines/mirage/pipeline_mirage.py]: I moved it to the top.

File: scripts/convert_mirage_to_diffusers.py (hunk 3 above)
Comment (DavidBert) [reply scripts/convert_mirage_to_diffusers.py]: I change it to come from a config instead.

File: scripts/convert_mirage_to_diffusers.py (hunk 4 above)
Comment (DavidBert) [reply scripts/convert_mirage_to_diffusers.py]: I removed all dependency to previous ref pipeline, this was a mistake.

File: docs/source/en/api/pipelines/mirage.md [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+Mirage checkpoints only store the transformer and scheduler weights locally. The VAE and text encoder are automatically loaded from HuggingFace during pipeline initialization:
+
+```py
+from diffusers import MiragePipeline
+
+# Load pipeline - VAE and text encoder will be loaded from HuggingFace
+pipe = MiragePipeline.from_pretrained("path/to/mirage_checkpoint")
+pipe.to("cuda")
+
+prompt = "A digital painting of a rusty, vintage tram on a sandy beach"
```
Comment (photoroman) [inline docs/source/en/api/pipelines/mirage.md]: We should use a more viral example prompt. Once we chose a viral name for the model, we can match the prompt to the name ;) Maybe we have to activate the Photoroom marketing team for this one.

File: docs/source/en/api/pipelines/mirage.md [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+
+## Loading the Pipeline
+
+Mirage checkpoints only store the transformer and scheduler weights locally. The VAE and text encoder are automatically loaded from HuggingFace during pipeline initialization:
+
+```py
+from diffusers import MiragePipeline
+
+# Load pipeline - VAE and text encoder will be loaded from HuggingFace
+pipe = MiragePipeline.from_pretrained("path/to/mirage_checkpoint")
```
Comment (photoroman) [inline docs/source/en/api/pipelines/mirage.md]: I guess we'll be able to store the checkpoint on Hugging Face as well, right? If yes, we should not forget to update the paths here to the official one, to make this truly copy-paste and run.

**Review by photoroman — APPROVED** [review]
LGTM!

EDIT: We should still discuss the name "Mirage" though before opening a PR to upstream.

**Review by photoroman — APPROVED** [review]
Had a quick look. Great! I trust Claude renamed everything correctly.

File: docs/source/en/api/pipelines/mirage.md [hunk 15]
```diff
@@ -46,7 +40,7 @@ from diffusers import MiragePipeline
 pipe = MiragePipeline.from_pretrained("path/to/mirage_checkpoint")
 pipe.to("cuda")
 
-prompt = "A digital painting of a rusty, vintage tram on a sandy beach"
+prompt = "A vibrant night sky filled with colorful fireworks, with one large firework burst forming the glowing text “Photon” in bright, sparkling light"
```
Comment (photoroman) [inline docs/source/en/api/pipelines/mirage.md]: Haha, awesome!
